### ISSUE TITLE
Simplify diagnostics story for custom ALC

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#51069]
### Description

When issues like [these](https://github.com/dotnet/aspnetcore/issues/31637) come up, it's a game of wack-a-mole to figure out what objects from a custom ALC are rooted by objects in the default ALC. Can we do something low tech to make this a bit easier? Maybe an SOS command.

### Regression?

No

cc @vitek-karas @janvorli

**Comment by jkotas** [comment]
> Can we do something low tech to make this a bit easier? Maybe an SOS command.

The sos command for this is gcroot. Here is an example for how it can be used to investigate leaking ALCs https://github.com/dotnet/runtime/issues/31377#issuecomment-549059548

Leaking ALCs are managed memory leaks, so most tutorials and tricks for debugging managed memory leaks are applicable too.

**Comment by davidfowl** [comment]
Yea, I know but the repro I have is extremely unreliable, so maybe it's a bug. I think would be interesting to be able to identify which ALC an object came from using SOS.

**Comment by davidfowl** [comment]
Oh that thread helped. So LoaderAllocator  is the magic I should be looking for.

**Comment by davidfowl** [comment]
Turns out event sources are really bad in this regard (probably all diagnostics really). It's not self contained enough to be unloadable.

**Comment by vitek-karas** [comment]
Just curious - do you think it happens to all event sources, or just those somebody is listening to? In theory we should be able to "do nothing" if nobody's listening... but maybe not.

Also - this should be fixable - we just need to make the references to event sources weak references (and handle the case where it goes away).

**Comment by davidfowl** [comment]
I hope we can solve it in general. EventSources basically register with etw or event pipe and end up getting finalized. Now it makes sense why I was seeing this repro be unreliable, at least when I was trying to narrow it down. I'll try to make a repro today

**Comment by janvorli** [comment]
@davidfowl are you aware of the following doc I have written on using and debugging unloadability? https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability#debug-unloading-issues

**Comment by davidfowl** [comment]
Nope! That's super useful! Also I figured out this issue, it's pretty cool one 😄

**Comment by vitek-karas** [comment]
@davidfowl Can you please create a separate issue for the EventSource unloadability? (Since you have the data/samples)

**Comment by davidfowl** [comment]
Turns out EventSources weren't the problem. Well I should say, they get finalized eventually. 

Here's the explanation https://github.com/dotnet/aspnetcore/issues/31637#issuecomment-817602892

TL;DR, async stack unwinding will root things (which makes sense). So extra care is needed when you execute async code in another load context. It's super easy to keep things alive.

**Comment by davidfowl** [comment]
We should add this one to the doc. Basically the equivalent of ExecuteAndUnload but the async version
